### PR TITLE
INTEL MKL: Enhance Mkl Relu Op with primitive reuse

### PR DESCRIPTION
**Pull request by gzmkl**
Enable MKL primitive reuse for Relu Op via caching, to improve
  (1) model training and
  (2) inference of small batch size
by minimizing primitive creation time.

************ Notes *******************
   Please review and merge this PR first 
          https://github.com/tensorflow/tensorflow/pull/19399

### CODE REVIEW

**Comment by gzmkl** [comment]
Close temporarily - pending conv_bwd PR

**Comment by gzmkl** [comment]
mkl_conv_ops.cc has been reverted to avoid any review confusion. 
The new version of mkl_conv_ops.cc is contained in PR #19399. 

Thanks

**Comment by drpngx** [comment]
@rmlarsen there are a few of these CLs, I am assigning them to you. You could try to enlist @tatianashp if she's up for it.

**Comment by gzmkl** [comment]
Pending on https://github.com/tensorflow/tensorflow/pull/19754

**Comment by gzmkl** [comment]
Reopen since PR #19399 has been merged

**Comment by yiqianglee** [comment]
Yes, i have already singed CLA, and should be ok to contribute.

**Comment by gzmkl** [comment]
Hi,
I am the PR submitter and confirm that I am OK with all the changes made by yiqianglee.
Thanks,
GZ

**Comment by rmlarsen** [comment]
@gzmkl I think many of my comments for PR 19403 would apply to this as well, please modify accordingly.

**Comment by gzmkl** [comment]
Plan to apply  PR 19403 comments on this PR too.

**Comment by rmlarsen** [comment]
@gzmkl resolved conflict.

**Comment by drpngx** [comment]
I think the Mac error is unrelated to this CL.

```
ERROR: /Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/external/com_google_googletest/BUILD.bazel:45:1: Couldn't build file external/com_google_googletest/_objs/gtest/external/com_google_googletest/googletest/src/gtest-death-test.o: C++ compilation of rule '@com_google_googletest//:gtest' failed: Unexpected IO error.: Running '/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/_bin/xcode-locator 8.3.3' failed with code 1.
This most likely indicates that xcode version 8.3.3 is not available on the host machine.
Process exited with status 1
stdout: error: Unable to extract CFBundleShortVersionString from URL: file:///Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/external/grpc/tools/dockerfile/distribtest/node_ubuntu1604_x64/
```

**Comment by rmlarsen** [comment]
@gzmkl can you ping this PR when you are done applying the changes suggested in PR 19403, please?

**Comment by gzmkl** [comment]
Hi Rasmus, I am done with code change with this PR. Thanks!

**Comment by gzmkl** [comment]
I have done code change per code review suggestions from other PR (19403). 

Please ignore changes in mkl_util.h after PR 19403/19402 have been approved.  

Thanks!

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/core/util/mkl_util.h [hunk 1]
```diff
@@ -1908,9 +1908,9 @@ class MklPrimitiveFactory {
   void SetOp(const string& key, MklPrimitive* op) {
     auto stream_iter = MklPrimitiveFactory<T>::GetHashMap().find(key);
 
-    CHECK(stream_iter == MklPrimitiveFactory<T>::GetHashMap().end());
+    CHECK(stream_iter == map.end());
```
[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/util/mkl_util.h]: It seems that mkl_util.h is screwed up, after quite a lot of code changes
per BatchNorm/Pooling primitive reuse PR. 

Let me check the source code.

[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/util/mkl_util.h]: I re-based mkl_util.h file with the public "master" branch. 
So the file in the PR branch is now same as the one in the master branch. 

Build/test shows no issue. 

Thanks

**Comment by gzmkl** [comment]
Hi,

I have done code change with PR (taking review comments of previous related PRs such as #19403; rebase mkl_util.h to avoid merge conflicts etc). Please let me know if I need to take future actions.

Thanks!

**Comment by drpngx** [comment]
Could you pull rebase and push again to resolve the conflicts? Thanks.

**Comment by drpngx** [comment]
@yiqianglee do you approve the commit for the CLA?

**Comment by gzmkl** [comment]
Hi, 
Ramesh of our team will take care of merge conflicts, as he has seen many of such on multiple PRs but did not have time to fix this PR.

Thanks,
GZ

**Comment by agramesh1** [comment]
I am ok with my commits being contributed to this project.

**Comment by gzmkl** [comment]
I am OK with the change (to address merge conflicts) by Ramesh.

**Comment by gzmkl** [comment]
Hi, 
Ramesh has helped to address the merge conflict. 
Somehow the "cla" tag still says "no". Please help to address the cla problem.

We would like to merge this PR asap, because we need to submit Relu3D before the end of the Friday, to give a two-week window for getting into Tensorflow 1.11.

Thank you!

**Comment by drpngx** [comment]
@rmlarsen if you have time, then the authors would like to have this reviewed in the next couple of days or so to make the `1.11` release cutoff.

**Comment by yiqianglee** [comment]
I am OK with the change by Ramesh and GZ.

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@gzmkl Thanks. Sorry about the review delay.

**Comment by gzmkl** [comment]
Rasmus,
Thank you very much for the diligent work of review all primitive reuse PRs. 
There were too much code change (almost rewriting MKL-DNN integration code). 
We truly appreciate your hard work with valuable code refinement suggestions. Great collaboration!
-GZ

**Comment by gzmkl** [comment]
Hi, 
I saw this PR approved but was not merged. There was a time-time during "Internal CI Build".
Please let me know what I can help. Thanks!  -GZ
